### ISSUE TITLE
GATv2 Implementation difference

### ISSUE THREAD

**Issue by alonsocampana** [pyg-team/pytorch_geometric#4184]
### 🛠 Proposed Refactor

Hello,
I was exploring the implementation of GATv2 and I noticed a difference with respect to the paper:
- In the paper the features, during message passing, the features of nodes, (and potentially edges)  are concatenated , and multiplied by a set of weights.
- In the implementation the features of the nodes are multiplied by a set of right and left weights, the same with edge features, and then are summed. 
Changing this would make the algorithm more expensive, but I'm wondering if without doing so, the algorithm is unable to compute dynamic attention, which is the intent of GATv2. I will try to benchmark the difference both in time and in performance.

### Suggest a potential alternative/fix

The change would be to pass a single set of weights:
``` python
self.lin = Linear(in_channels*2+edge_size, heads * out_channels, bias=bias,
                                weight_initializer='glorot')
```

and then update the message passing to deal with it:

``` python
        if edge_attr is not None:
            x = torch.cat([x_j, x_i, edge_attr], dim=1)
        else:
            x = torch.cat([x_j, x_i], dim=1)
        x = self.lin(x).view(-1, self.heads, self.out_channels)
        x_ = F.leaky_relu(x, self.negative_slope)
        alpha = (x_ * self.att).sum(dim=-1)
        alpha = softmax(alpha, index, ptr, size_i)
        self._alpha = alpha
        alpha = F.dropout(alpha, p=self.dropout, training=self.training)
        return x * alpha.unsqueeze(-1)
```

EDIT:
I did a very naive test using cora, without further linear layers, and the loss is consistently ~15% lower. The execution time is +~150% higher.

**Comment by rusty1s** [comment]
I'm pretty sure our implementation matches with the one in the paper, so I appreciate it if you could resolve my doubts. From a computation point perspective, this looks equivalent (we use different transformations for source and destination node features and concatenate them rather than concatenate first and apply a big transformation matrix). The expressive power of `GATv2` is still maintained due to the non-linearity before applying attention weights.

**Comment by alonsocampana** [comment]
You are right. W_l * h_i  + W_r * h_j is equivalent to W*[h_i || h_j], the difference in runtime is obvious (lots of additional multiplications). Wondering if the difference is performance is due to a different computational graph, or some difference from my side.

**Comment by rusty1s** [comment]
It might be a good sanity check to check for equivalence in output given the same weight initializations.
